cEP-0025.md

```diff
@@ -0,0 +1,90 @@
+pyflakes-enhanced-AST into coala.
+
+## Introduction
+
+The current approach of plugin development for flake8 does not utilises 
+the full potential of enhanced AST rather they simply depend on python AST 
+and so, a whole lot of rework is required to do the basic traversing
+and collection of important nodes.
+
+Pyflakes provides with a basic API that does the traversing.
+So, if a developer uses enhanced AST he just needs to work on 
+the implementation of the new logic that his/her plugin provides 
+and not about the fidelity of the basic node handlers.
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
+2. The repository will contain a new package `pyflakes-plugins` which
+   would house all bears which uses pyflakes-enhanced-ast.
+3. The `pyflakes-plugins` would contain a metabear `PyflakesASTBear`
+   which would be used by all other plugin bears.
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being 
+used in `coala-bears` to test all its plugins.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations of PyflakesASTBear:
+
+```python
+import ast
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.results.HiddenResult import HiddenResult
+from pyflakes.checker import Checker
+from pyflakes.checker import ClassScope, FunctionScope, ModuleScope
+from pyflakes.checker import GeneratorScope, DoctestScope
+
+
+class PyflakesASTBear(LocalBear):
+    def run(self, filename, file):
+        """
+            Generates pyflakes-enhance-AST for the input file and returns
+            deadScopes as HiddenResult
+            :return:
+                One HiddenResult containing a dictionary with keys being
+                type of scope and values being a list of scopes generated
+                from the file.
+        """
+        tree = ast.parse(''.join(file))
+        result = Checker(tree, filename)
+
+        def get_scope(scope_level, scopes):
+            return list(filter(lambda scope: isinstance(scope, scope_level),
+                               scopes))
+
+        content = {
+                    'ClassScope': get_scope(ClassScope,
+                                            result.deadScopes),
+                    'FunctionScope': get_scope(FunctionScope,
+                                               result.deadScopes),
+                    'ModuleScope': get_scope(ModuleScope,
+                                             result.deadScopes),
+                    'GeneratorScope': get_scope(GeneratorScope,
+                                                result.deadScopes),
+                    'DoctestScope': get_scope(DoctestScope,
+                                              result.deadScopes)
+        }
+
+        yield HiddenResult(self, content)
```

Comment:
create a HiddenResult subclass, and use OO. a dict is ugly, and brittle.  backwards compatibility and deprecation with a dict is limited, so a dict will mean we have problems when the pyflakes internals change.

Also please see all the issues related to HiddenResults.  Fixing a few of them in your project will make everything much neater.